pkg/updatestatus/nodeinformer.go

```diff
@@ -0,0 +1,390 @@
+		poolsMap[pool.Name] = pool
+		s, err := metav1.LabelSelectorAsSelector(pool.Spec.NodeSelector)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get label selector from the pool %s: %w", pool.Name, err)
+		}
+		switch pool.Name {
+		case mco.MachineConfigPoolMaster:
+			masterSelector = s
+		case mco.MachineConfigPoolWorker:
+			workerSelector = s
+		default:
+			customSelectors[pool.Name] = s
+		}
+	}
+
+	if masterSelector != nil && masterSelector.Matches(labels.Set(node.Labels)) {
+		return poolsMap[mco.MachineConfigPoolMaster], nil
+	}
+	for name, selector := range customSelectors {
+		if selector.Matches(labels.Set(node.Labels)) {
+			return poolsMap[name], nil
+		}
+	}
+	if workerSelector != nil && workerSelector.Matches(labels.Set(node.Labels)) {
+		return poolsMap[mco.MachineConfigPoolWorker], nil
+	}
+	return nil, fmt.Errorf("failed to find a matching node selector from %d machine config pools", len(pools))
+}
+
+func isNodeDegraded(node *corev1.Node) bool {
+	// Inspired by: https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/node/status.go
+	if node.Annotations == nil {
+		return false
+	}
+	dconfig, ok := node.Annotations[mco.DesiredMachineConfigAnnotationKey]
+	if !ok || dconfig == "" {
+		return false
+	}
+	dstate, ok := node.Annotations[mco.MachineConfigDaemonStateAnnotationKey]
+	if !ok || dstate == "" {
+		return false
+	}
+
+	if dstate == mco.MachineConfigDaemonStateDegraded || dstate == mco.MachineConfigDaemonStateUnreconcilable {
+		return true
+	}
+	return false
+}
+
+func isNodeDraining(node *corev1.Node, isUpdating bool) bool {
+	desiredDrain := node.Annotations[mco.DesiredDrainerAnnotationKey]
+	appliedDrain := node.Annotations[mco.LastAppliedDrainerAnnotationKey]
+
+	if appliedDrain == "" || desiredDrain == "" {
+		return false
+	}
+
+	if desiredDrain != appliedDrain {
+		desiredVerb := strings.Split(desiredDrain, "-")[0]
+		if desiredVerb == mco.DrainerStateDrain {
+			return true
+		}
+	}
+
+	// Node is supposed to be updating but MCD hasn't had the time to update
+	// its state from original `Done` to `Working` and start the drain process.
+	// Default to drain process so that we don't report completed.
+	mcdState := node.Annotations[mco.MachineConfigDaemonStateAnnotationKey]
+	return isUpdating && mcdState == mco.MachineConfigDaemonStateDone
+}
+
+func determineConditions(pool *machineconfigv1.MachineConfigPool, node *corev1.Node, isUpdating, isUpdated, isUnavailable, isDegraded bool, lns *mco.LayeredNodeState, now metav1.Time) ([]metav1.Condition, string, *metav1.Duration) {
+	var estimate *metav1.Duration
+	var message string
+
+	updating := metav1.Condition{
+		Type:               string(NodeStatusInsightUpdating),
+		Status:             metav1.ConditionUnknown,
+		Reason:             string(NodeCannotDetermine),
+		LastTransitionTime: now,
+	}
+	available := metav1.Condition{
+		Type:               string(NodeStatusInsightAvailable),
+		Status:             metav1.ConditionTrue,
```

Comment:
`Reason` is required in `metav1.Condition` so we need it for positive cases, too. `AsExpected` is commonly used in such case.

Same for `Message`. The message can be something very brief for "boring" case but needs to be there.